stl/inc/vector

```diff
@@ -458,6 +494,139 @@ public:
+        auto& _My_data = _Myvec_._Mypair._Myval2;
+        _Apply_annotation(_My_data._Myfirst, _My_data._Myend, _My_data._Mylast, _My_data._Myend);
+    }
+
+    static _CONSTEXPR20 void _Change_annotation(const vector& _Myvec_, const difference_type _Count) noexcept {
+        // Extends /shrinks the annotated range by _Count
+        auto& _My_data = _Myvec_._Mypair._Myval2;
+        _Apply_annotation(_My_data._Myfirst, _My_data._Myend, _My_data._Mylast, _My_data._Mylast + _Count);
+    }
+
+    static const void* _Get_aligned_first(const void* _First, const void* _End) noexcept {
+        const char* _CFirst    = reinterpret_cast<const char*>(_First);
+        const char* _CEnd      = reinterpret_cast<const char*>(_End);
+        const size_t _Capacity = static_cast<size_t>(_CEnd - _CFirst);
+
+        if (_Capacity >= 8) {
+            // We are guaranteed to have sufficient space to find an aligned address.
+            return reinterpret_cast<void*>((reinterpret_cast<uintptr_t>(_CFirst) + 7) & ~7);
+        }
+
+        uintptr_t _Alignment_offset = reinterpret_cast<uintptr_t>(_CFirst) & 7;
+        if (_Alignment_offset) {
+            _Alignment_offset = 8 - _Alignment_offset;
+        }
+
+        if (_Capacity > _Alignment_offset) {
+            return _CFirst + _Alignment_offset;
+        }
+
+        return nullptr;
+    }
+
+    static _CONSTEXPR20 void _Apply_annotation(
+        pointer _First_, pointer _End_, pointer _Old_last_, pointer _New_last_) noexcept {
+#if _HAS_CXX20
+        if (_STD is_constant_evaluated()) {
+            return;
+        }
+#endif // _HAS_CXX20
+
+        if (!_Asan_vector_should_annotate) {
+            return;
+        }
+
+        const char* _First    = reinterpret_cast<const char*>(_Unfancy_maybe_null(_First_));
+        const char* _End      = reinterpret_cast<const char*>(_Unfancy_maybe_null(_End_));
+        const char* _Old_last = reinterpret_cast<const char*>(_Unfancy_maybe_null(_Old_last_));
+        const char* _New_last = reinterpret_cast<const char*>(_Unfancy_maybe_null(_New_last_));
+        if constexpr (_Has_minimum_allocation_alignment<vector>) {
+            __sanitizer_annotate_contiguous_container(_First, _End, _Old_last, _New_last);
+        } else {
+            const void* _Aligned_first = _Get_aligned_first(_First, _End);
+            if (!_Aligned_first) {
+                // There is no aligned address within the underlying buffer; nothing to do.
+                return;
+            }
+
+            const void* _Aligned_old_last = _Old_last < _Aligned_first ? _Aligned_first : _Old_last;
+            const void* _Aligned_new_last = _New_last < _Aligned_first ? _Aligned_first : _New_last;
+            const void* _Aligned_end      = _End < _Aligned_first ? _Aligned_first : _End;
+            __sanitizer_annotate_contiguous_container(
+                _Aligned_first, _Aligned_end, _Aligned_old_last, _Aligned_new_last);

```

Comment:
I'm uncertain whether this could cause false positive detections (and @CaseyCarter agrees after looking at it again). If we're incrementing to get `_Aligned_first`, then it seems like we're telling ASAN that a smaller region of the vector's allocated-and-constructed elements are actually valid, so uses of the vector's very first elements would fall outside of the range we told ASAN was valid.

It seems that if we're handling an allocator that's returning insufficiently aligned memory, that we should be *decreasing* the first, which would be telling ASAN that a slightly larger region of memory is valid - i.e. risking false negatives (any dereferencing of the -1, -2, etc. positions), but with no risk of false positives.

If we're confused, let's talk about this further (in comments or on Discord).

Reply:
Starting the region at `_Aligned_first` doesn't cause the runtime to annotate the region in front of `_Aligned_first` as invalid. You can basically think of `__sanitizer_annotate_contiguous_container` as a directive to `memset` some shadow memory with some _very_ basic checks using the beginning and the end parameters.
See: https://github.com/llvm/llvm-project/blob/de3038400b163d91dca743210ce9b10001411208/compiler-rt/lib/asan/asan_poisoning.cpp#L343

Reply:
Starting the region at `_Aligned_first` doesn't cause the runtime to annotate the region in front of `_Aligned_first` as invalid. You can basically think of `__sanitizer_annotate_contiguous_container` as a directive to `memset` some shadow memory with some _very_ basic checks using the beginning and the end parameters.
See: https://github.com/llvm/llvm-project/blob/de3038400b163d91dca743210ce9b10001411208/compiler-rt/lib/asan/asan_poisoning.cpp#L343

Reply:
Ok, thanks.